### PR TITLE
refactor: replace any types with specific types

### PR DESCRIPTION
**Pull request by ChristianMurphy**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [ ] tests are included
- [ ] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

follow up to #2186 

[contributor license agreement]: https://www.apereo.org/node/676
[commit guidelines]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: Source/Plugins/Core/com.equella.core/js/tsrc/cloudcontrol/index.ts [hunk 1]
```diff
@@ -76,6 +76,7 @@ interface CommandsPromise {
 }
 
 interface Registration {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
   mount: (api: ControlApi<any>) => void;
```
Comment (ChristianMurphy) [inline Source/Plugins/Core/com.equella.core/js/tsrc/cloudcontrol/index.ts:80]: https://github.com/openequella/openEQUELLA/pull/2008#pullrequestreview-443552493

File: Source/Plugins/Core/com.equella.core/js/tsrc/legacycontent/LegacyContent.tsx [hunk 2]
```diff
@@ -23,12 +23,10 @@ import { API_BASE_URL } from "../config";
 
 declare global {
   interface Window {
-    _trigger: any;
```
Comment (ChristianMurphy) [inline Source/Plugins/Core/com.equella.core/js/tsrc/legacycontent/LegacyContent.tsx:26]: never used, so removed

File: Source/Plugins/Core/com.equella.core/js/tsrc/legacycontent/LegacyContent.tsx [hunk 3]
```diff
@@ -23,12 +23,10 @@ import { API_BASE_URL } from "../config";
...
-    eval: any;
```
Comment (ChristianMurphy) [inline Source/Plugins/Core/com.equella.core/js/tsrc/legacycontent/LegacyContent.tsx:27]: this is already a global, so allowed this to go back to the official typing

File: Source/Plugins/Core/com.equella.core/js/tsrc/legacycontent/LegacyContent.tsx [hunk 4]
```diff
@@ -300,8 +298,8 @@ function updateStylesheets(
       newCss.rel = "stylesheet";
       newCss.href = cssUrl;
       head.insertBefore(newCss, insertPoint);
-      const p = new Promise((resolve, reject) => {
-        newCss.addEventListener("load", resolve, false);
+      const p = new Promise<void>((resolve, reject) => {
+        newCss.addEventListener("load", () => resolve(), false);
```
Comment (ChristianMurphy) [inline Source/Plugins/Core/com.equella.core/js/tsrc/legacycontent/LegacyContent.tsx:302]: Promise is only used to detect completion, not a value.
This is now a `void` type, typescript chokes if a value is passed to void, so updated this to guarantee void.

File: Source/Plugins/Core/com.equella.core/js/tsrc/modules/SchemaModule.tsx [hunk 5]
```diff
@@ -85,15 +109,10 @@ export const buildSchemaTree = (
...
-      if (elements.length !== 1 && elements[0] !== standardRoot) {
-        throw new Error(
-          "Received schema does not start with the standard <xml> root element."
-        );
-      }
-      return buildSchemaTree(schema.definition[standardRoot], standardRoot);
+    ({ definition }: OEQ.Schema.EquellaSchema) => {
+      assertStandardRoot(definition);
+
+      return buildSchemaTree(definition[standardRoot], standardRoot);
```
Comment (ChristianMurphy) [inline Source/Plugins/Core/com.equella.core/js/tsrc/modules/SchemaModule.tsx:115]: buildSchemaTree needs an object, the assertion is added to typescript understands that definition is a `Definition`, not `unknown`.

File: Source/Plugins/Core/com.equella.core/js/tsrc/modules/SchemaModule.tsx [hunk 6]
```diff
@@ -59,20 +66,37 @@ export const schemaListSummary = async (): Promise<Map<string, string>> => {
...
 };
 
+const standardRoot = "xml";
+
+/**
+ * Assert that a definition is a standard root
+ */
+function assertStandardRoot(
+  value: Definition
+): asserts value is { xml: Definition } {
```
Comment (ChristianMurphy) [inline Source/Plugins/Core/com.equella.core/js/tsrc/modules/SchemaModule.tsx:90]: same code as before, wrapped in https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions

File: Source/Plugins/Core/com.equella.core/js/tsrc/modules/SchemaModule.tsx [hunk 7]
```diff
@@ -59,20 +66,37 @@ export const schemaListSummary = async (): Promise<Map<string, string>> => {
...
 ): SchemaNode => {
   const node: SchemaNode = { name: name, parent: parent };
-  node.children = Object.keys(definition)
-    .filter((childName: string) => typeof definition[childName] === "object")
-    .map((childName: string) =>
-      buildSchemaTree(definition[childName], childName, node)
+  node.children = Object.entries(definition)
+    .filter((pair): pair is [string, Definition] => typeof pair[1] === "object")
+    .map(([childName, childNode]) =>
+      buildSchemaTree(childNode, childName, node)
```
Comment (ChristianMurphy) [inline Source/Plugins/Core/com.equella.core/js/tsrc/modules/SchemaModule.tsx:77]: refactored, so the `filter` can be a predicate https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-type-predicates
narrowing the type for the `map` statement

**Comment by ChristianMurphy** [comment]
@edalex-ian codebuild seems unhappy, is that related to this change? or is there some flakiness?

**Comment by PenghaiZhang** [comment]
> @edalex-ian codebuild seems unhappy, is that related to this change? or is there some flakiness?

62 tests failed so I believe something is broken

**Comment by edalex-ian** [comment]
Yup, like @PenghaiZhang said, something is actually broken here.

```
Failed: Total 602, Failed 62, Errors 0, Passed 525, Skipped 15
```

Now that's a case of running the traditional Selenium tests with the New UI turned on. AWS CodeBuild runs in the same way - hence why that's also fallen over too. The issues will be something to do with either your work in `tsrc/legacycontent` and/or `tsrc/mainui` as that's wrapping all the 'legacy' content.

File: Source/Plugins/Core/com.equella.core/js/tsrc/mainui/routes.tsx [hunk 8]
```diff
@@ -29,24 +29,29 @@ import LoginNoticeConfigPage from "../loginnotice/LoginNoticeConfigPage";
...
   strict?: boolean;
-  to?: any;
+  to?: string | To;
+}
+
+export interface Routes {
+  [route: string]: OEQRoute<TemplateUpdateProps>;
 }
 
 export const routes = {
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/mainui/routes.tsx:57]: With the new `Routes` type, should this then be:

```suggestion
export const routes:Routes = {
```

Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/mainui/routes.tsx:57]: So with the new `Routes` type, should we now make this a bit more explicit with:

```suggestion
export const routes: Routes = {
```

**Review by edalex-ian — COMMENTED** [review]
I'll come and have another look when you get the build green. Obviously playing with legacy content etc is a bit risky, but by the same token we benefit by ensuring that's as robust as can be. (There's also some refactoring in general to be had there.)

For now then, I only have the one minor query.

**Pull request closed without merging**